scalding-core/src/main/scala/com/twitter/scalding/ExecutionOptimizationRules.scala

```diff
@@ -167,250 +184,69 @@ object ExecutionOptimizationRules {
-            case Single(w) => acc + isW(w)
-            case Many(h, rest, _) => loop(rest, acc + isW(h))
-            case MapZip(of, _) => loop(of, acc)
-          }
-
-        loop(this, 0)
-      }
-
-      /**
-       * The type of this ensures that the head of the HList
-       * must be a WriteExecution. Additionally, none of
-       * the other items are, but we don't prove that in the types,
-       * instead we prove that with tests
-       *
-       * Note, one might imagine we could always return a FlattenedZip
-       * here, if the optimization does not apply, just return `this`.
-       * That intuition is not quite correct: the type changes in the result
-       * Since we start as FlattenedZip[Ex, A] for some Ex[A] subtype of
-       * Execution[A] but we return
-       * FlattenedZip[WriteExecution, A], if there are no WriteExecution
-       * instances in the graph, we have no way of returning a FlattenedZip
-       * with one in it.
-       *
-       * It's a bit confusing because below when we use it in our rule,
-       * we only call it when we know it applies since we count
-       * the writes and verify there are 2 or more, but internally
-       * we recursively call this method, and we do exercise the None
-       * branch.
-       */
-      def optimize: Option[FlattenedZip[WriteExecution, A]]
-    }
+    private sealed trait FlattenedZip[+A]
 
     private object FlattenedZip {
-
-      /**
-       * Here is the simplest FlattenZip: just a single Execution
-       */
-      final case class Single[+Ex[x] <: Execution[x], A](ex: Ex[A]) extends FlattenedZip[Ex, A] {
-        type ExH = Execution[A]
-        def executions = ex
-        def toExecution(exa: Execution[A]): Execution[A] =
-          exa
-
-        def zip[B](that: FlattenedZip[Execution, B]) =
-          that match {
-            case s@Single(ex2) => Many[Execution, A, B](ex, s)
-            case m@Many(_, _, _) =>
-              def go[X, Y](m: Many[Execution, X, Y, B]): FlattenedZip[Execution, (A, B)] = {
-                val sub0: SubTypes[(X, Y), B] = m.sub
-                val flat0: FlattenedZip[Execution, (A, (X, Y))] =
-                  Many[Execution, A, (X, Y)](ex, Single(m.head).zip(m.rest))
-                type F[+X] = FlattenedZip[Execution, (A, X)]
-                sub0.liftCo[F](flat0)
-              }
-              go(m)
-            case MapZip(left, fn) =>
-              zip(left).map(ZipMap.MapRight(fn))
-          }
-        def optimize: Option[FlattenedZip[WriteExecution, A]] =
-          ex match {
-            case w@WriteExecution(_, _, _) => Some(Single(w))
-            case _ => None
-          }
+      final case class NonWrite[T](nonWrite: Execution[T]) extends FlattenedZip[T]
+      final case class Write[T](write: WriteExecution[T]) extends FlattenedZip[T]
+      final case class Composed[T1, T2, T](write: WriteExecution[T1], nonWrite: Execution[T2], compose: (((T1, T2)) => T)) extends FlattenedZip[T]
```

Comment:
Why the extra parens around compose? Can we remove?

Reply:
Changed to Function1